atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/monitoring/TimestampTrackerTest.java

```diff
@@ -0,0 +1,161 @@
+        }
+    }
+
+    @Test
+    public void metricsAreDeregisteredUponClose() {
+        try (TimestampTracker tracker = new TimestampTracker(Clock.defaultClock())) {
+            tracker.registerTimestampForTracking(FAKE_METRIC, () -> 1L);
+            assertThat(AtlasDbMetrics.getMetricRegistry().getNames())
+                    .contains(buildFullyQualifiedMetricName(FAKE_METRIC));
+        }
+
+        assertThat(AtlasDbMetrics.getMetricRegistry().getNames())
+                .doesNotContain(buildFullyQualifiedMetricName(FAKE_METRIC));
+    }
+
+    @Test
+    public void canCloseMultipleTimes() {
+        TimestampTracker tracker = new TimestampTracker(Clock.defaultClock());
+        tracker.registerTimestampForTracking(FAKE_METRIC, () -> 1L);
+
+        tracker.close();
+        tracker.close();
+        assertThat(AtlasDbMetrics.getMetricRegistry().getNames())
+                .doesNotContain(buildFullyQualifiedMetricName(FAKE_METRIC));
+    }
+
+    @Test
+    public void doesNotCallSupplierOnRequestsWithinRetriggerInterval() {
+        try (TimestampTracker tracker = new TimestampTracker(mockClock)) {
+            when(mockClock.getTick()).thenReturn(0L, 1L, 2L);
+            AtomicLong timestampValue = new AtomicLong(0L);
+            tracker.registerTimestampForTracking(FAKE_METRIC, timestampValue::incrementAndGet);
+
+            assertThat(getGauge(FAKE_METRIC).getValue()).isEqualTo(1L);
+            assertThat(getGauge(FAKE_METRIC).getValue()).isEqualTo(1L);
+            assertThat(getGauge(FAKE_METRIC).getValue()).isEqualTo(1L);
+        }
+    }
+
+    @Test
+    public void callsSupplierAgainAfterTimeElapses() {
+        try (TimestampTracker tracker = new TimestampTracker(mockClock)) {
+            when(mockClock.getTick()).thenReturn(0L, CACHE_RETRIGGER_NANOS / 2, CACHE_RETRIGGER_NANOS);
+            AtomicLong timestampValue = new AtomicLong(0L);
+            tracker.registerTimestampForTracking(FAKE_METRIC, timestampValue::incrementAndGet);
+
+            assertThat(getGauge(FAKE_METRIC).getValue()).isEqualTo(1L);
+            assertThat(getGauge(FAKE_METRIC).getValue()).isEqualTo(1L);
+            assertThat(getGauge(FAKE_METRIC).getValue()).isEqualTo(2L);
+        }
+    }
+
+    @Test
+    public void doesNotCallSupplierUnlessGaugeIsQueried() {
+        try (TimestampTracker tracker = new TimestampTracker(mockClock)) {
+            when(mockClock.getTick()).thenReturn(0L, CACHE_RETRIGGER_NANOS * 50000);
+            AtomicLong timestampValue = new AtomicLong(0L);
+            tracker.registerTimestampForTracking(FAKE_METRIC, timestampValue::incrementAndGet);
+
+            assertThat(getGauge(FAKE_METRIC).getValue()).isEqualTo(1L);
+            assertThat(getGauge(FAKE_METRIC).getValue()).isEqualTo(2L);
+        }
+    }
+
+    private static String buildFullyQualifiedMetricName(String shortName) {
+        return "com.palantir.atlasdb.monitoring.TimestampTracker." + shortName;
```

Comment:
nit: this is somewhat sensitive to the `TimestampTracker` being moved to another package.

Reply:
Replaced with `MetricRegistry.name(TimestampTracker.class, shortName);`, which should handle the class name (and even changes in how codahale metrics creates its names) correctly.